### PR TITLE
SDKS-965 Support Apple SignIn

### PR DESCRIPTION
**Pull request by witrisna**

### CODE REVIEW

**Review by spetrov — CHANGES_REQUESTED** [review]
Looks good to me! 👍 
I have a few minor questions. Please take a look.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AuthServiceClient.java [hunk 1]
```diff
@@ -9,12 +9,14 @@
 
 import android.net.Uri;
 
+import org.forgerock.android.auth.callback.AdditionalParameterCallback;
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/AuthServiceClient.java:12]: Update copyright

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                appAuthConfigurationBuilder.build());
+
+        CustomTabsIntent.Builder intentBuilder = getIntentBuilder(authorizationService, authorizationRequest.toUri());
+
+        Intent intent = authorizationService.getAuthorizationRequestIntent(
+                authorizationRequest, intentBuilder.build());
+        startActivityForResult(intent, RC_SIGN_IN);
+    }
+
+    public CustomTabsIntent.Builder getIntentBuilder(AuthorizationService service, Uri uri) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java:84]: javadoc?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/FacebookSignInHandler.java [hunk 3]
```diff
@@ -79,22 +78,22 @@ public void onActivityResult(int requestCode, int resultCode, @Nullable Intent d
 
     @Override
     public String getTokenType() {
-        return "access_token";
+        return ACCESS_TOKEN;
     }
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/idp/FacebookSignInHandler.java:84]: extra line

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/IdPResult.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * SignIn Result from the {@link IdPHandler}
+ */
+@Getter
+public class IdPResult {
+
+    private final String token;
+    private final Map<String, String> additionalParameters;
+
+    public IdPResult(String token, Map<String, String> additionalParameters) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/idp/IdPResult.java:23]: javadoc?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/IdPResult.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        this.token = token;
+        this.additionalParameters = additionalParameters;
+    }
+
+    public IdPResult(String token) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/idp/IdPResult.java:28]: javadoc?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AuthServiceClient.java (hunk 1 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/AuthServiceClient.java:12]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+
+        AuthorizationRequest.Builder authRequestBuilder = new AuthorizationRequest.Builder(configuration,
+                idPClient.getClientId(),
+                CODE,
+                Uri.parse(idPClient.getRedirectUri()))
+                .setAdditionalParameters(Collections.singletonMap("nonce", idPClient.getNonce()))
+                .setScopes(idPClient.getScopes())
+                .setState(null)
+                .setResponseMode(FORM_POST);
```
[the remark this answers is not in the document]
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java:70]: Similar to above issue, that attribute does not include in the IdPCallback

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+/**
+ * {@link IdPHandler} to handle Apple login
+ */
+public class AppleSignInHandler extends Fragment implements IdPHandler {
+
+    public static final int RC_SIGN_IN = 1000;
+    public static final String TAG = AppleSignInHandler.class.getName();
+    public static final String FORM_POST_ENTRY = "form_post_entry";
+    public static final String AUTHORIZE_ENDPOINT = "https://appleid.apple.com/auth/authorize";
+    public static final String TOKEN_ENDPOINT = "https://appleid.apple.com/auth/token";
```
[the remark this answers is not in the document]
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java:46]: Currently, the IdPCallback does not include the authorization and token endpoint. That may need to change the server to return those values, another option is to use well-known endpoint. I would keep it hardcode for now.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+/**
+ * {@link IdPHandler} to handle Apple login
+ */
+public class AppleSignInHandler extends Fragment implements IdPHandler {
+
+    public static final int RC_SIGN_IN = 1000;
+    public static final String TAG = AppleSignInHandler.class.getName();
+    public static final String FORM_POST_ENTRY = "form_post_entry";
+    public static final String AUTHORIZE_ENDPOINT = "https://appleid.apple.com/auth/authorize";
```
[the remark this answers is not in the document]
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java:45]: Currently, the IdPCallback does not include the authorization and token endpoint. That may need to change the server to return those values, another option is to use well-known endpoint. I would keep it hardcode for now.

[the remark this answers is not in the document]
Comment (spetrov) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java:45]: sounds good! 👍

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (spetrov) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/idp/AppleSignInHandler.java:70]: 👍

**Review by spetrov — APPROVED** [review]
Looks good to me! 👍 
Approved!

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/IdPResult.java (hunk 4 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/idp/IdPResult.java:30]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/idp/IdPResult.java (hunk 5 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/idp/IdPResult.java:40]: Updated

**Review by rodrigoareis — APPROVED** [review]
Looks good to me

**Pull request merged**